### ISSUE TITLE
Process: Avoid deadlock when standard streams are not redirected to files

### ISSUE THREAD

**Issue by jonilam** [robotframework/robotframework#4173]
One of our test cases started to hang after switching from Run And Return Rc And Output-keyword to a Run Process-keyword.  We debugged the issue and noticed that one of the threads of our test application was stuck in fflush(). Issue is caused by the usage of  Popen.wait() + read in Process-library. There is a note about this in a python subprocess-documentation:

"This will deadlock when using stdout=PIPE or stderr=PIPE and the child process generates enough output to a pipe such that it blocks waiting for the OS pipe buffer to accept more data. Use Popen.communicate() when using pipes to avoid that."
https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait

As stated by the note above, this happens only when pipes are used, and can be avoided by writing stdout/stderr to a file. I verified that indeed using Popen.communicate() instead of Popen.wait() fixes the problem.

Noticed the issue with RF 3.1.1 but tested that the issue is still there in version 4.1.3.

**Comment by pekkaklarck** [comment]
I don't remember why exactly we use `wait` and not `communicate`. We possibly could change that if the user hasn't provided `stdout` and `stderr`. That's the default so it could avoid hanging also in other cases.

Did you test does redirecting `stdout/err` to a file fox the problem? If it does, that's a good workaround.

**Comment by jonilam** [comment]
Yes, we are currently using stdout/err redirecting to a file as a workaround and it is working without issues.

**Comment by pekkaklarck** [comment]
Good to know that redirecting to files works. It would still be good to investigate could we use `communicate` when `stdout` and `stderr` are PIPEs (which is the default) to avoid deadlocking in cases like this. A problem is that although we call `wait()`, we only do that after elsewhere calling `poll()` to verify that the process has stopped and my assumption is that it hangs already there. It's thus possible we needed to restructure the code a bit more. Good thing is that we have lot of tests so even bigger changes can be done safely.

Unfortunately I'm super busy and don't have time to look at this in the near future. I leave this issue open in case someone else is interested to look at it.

**Comment by pekkaklarck** [comment]
PR #5302 implements this.

**Comment by franzhaas** [comment]
> PR #5302 implements this.

@pekkaklarck can we put this into the 7.3 scope?

I am sorry but I failed to do this myself

**Comment by franzhaas** [comment]
this is also related to this issue.: https://github.com/robotframework/robotframework/issues/5345

**Comment by pekkaklarck** [comment]
I fixed this in f2f2f99ad43825162109ae83e7e6886bcdfa60ba. The fix is based on PR #5302 by @franzhaas, but contains some cleanup, doc updates, and fix for handling closed stdout/stderr PIPEs. Big thanks for the PR!
